Make BlockingOperatorToIterator exert backpressure.

vqvu opened the pull request:
The iterator created by `BlockingOperator#getIterator()` doesn't exert backpressure, which causes code like this to never terminate/run out of memory

```java
Observable.from(new Iterable<Integer>() {

    @Override
    public Iterator<Integer> iterator() {
        return new Iterator<Integer>() {

            @Override
            public boolean hasNext() {
                return true;
            }

            @Override
            public Integer next() {
                return 1;
            }
        };
    }
}).toBlocking().getIterator().next();
```

This PR adds the appropriate `request` calls so that this works. I had to combine the implementations of `Subscriber` and `Iterator` into a single class to get access to `request()`.

akarnokd commented on src/main/java/rx/internal/operators/BlockingOperatorToIterator.java:
```diff
@@ -47,68 +46,82 @@ private BlockingOperatorToIterator() {
-            private Notification<? extends T> buf;
+        public SubscriberIterator() {
+            this.notifications = new LinkedBlockingQueue<Notification<? extends T>>();
+            this.buf = null;
```
Setting variables to their default value on construction is a no-op semantically but still mean instructions executed which causes the construction of the object takes longer.

vqvu replied:
I'm surprised the compiler doesn't just optimize that out...but I can remove it.

akarnokd commented on src/main/java/rx/internal/operators/BlockingOperatorToIterator.java:
```diff
@@ -47,68 +46,82 @@ private BlockingOperatorToIterator() {
+        @Override
+        public boolean hasNext() {
+            if (buf == null) {
+                request(1);
```
The problem with this approach is that there can't be any cross-thread pipelining because one thread won't progress unless the other thread does 1 thing. A better approach would be to request in bigger batches and re-request after a certain proportion of the batch has been received.

vqvu replied:
This is equivalent to simply calling `request(batchSize)` in `onStart` right? Also, what batch size would you recommend?

akarnokd replied:
Right. We use `RxRingBuffer.SIZE` by default and the re-request value is 25% of it. [Here](https://github.com/ReactiveX/RxJava/blob/1.x/src/main/java/rx/internal/operators/OperatorMerge.java#L783) is an example for it.

vqvu commented:
Ok. I removed the unnecessary variable initialization and added batched requests. The initial batch is `RxRingBuffer.SIZE`, and subsequent ones are `3 * RxRingBuffer.SIZE / 4` each (once we've received that much).

akarnokd commented:
Excellent! :+1:

abersnaze commented:
:+1:
